src/channel.rs

```diff
@@ -442,6 +442,30 @@ impl ChannelBuilder {
         let channel =
             unsafe { grpc_sys::grpc_insecure_channel_create(addr_ptr, args.args, ptr::null_mut()) };
 
+        unsafe { self.connect_from_raw_channel(channel) }
+    }
+
+    /// Build an insecure [`Channel`] taking over an established connection from
+    /// a file descriptor. The target string given is purely informative to
+    /// describe the endpoint of the connection. Takes ownership of the given
+    /// file descriptor and will close it when the connection is closed.
+    ///
+    /// This function is available on posix systems only.
+    #[cfg(unix)]
+    pub unsafe fn connect_from_fd(mut self, target: &str, fd: ::std::os::raw::c_int) -> Channel {
+        let args = self.prepare_connect_args();
+        let target = CString::new(target).unwrap();
+        let target_ptr = target.as_ptr();
+        let channel = grpc_sys::grpc_insecure_channel_create_from_fd(target_ptr, fd, args.args);
+
+        self.connect_from_raw_channel(channel)
+    }
+
+    /// Build a channel from a raw underlying core grpc channel. This allows to
+    /// establish an RPC connection using custom channel implementations.
+    /// Takes over ownership of the given channel object and will close it
+    /// when the connection is closed.
+    pub unsafe fn connect_from_raw_channel(self, channel: *mut grpc_channel) -> Channel {
```

Comment:
You can mark `Channel::new` as public and unsafe instead.

Reply:
> You can mark `Channel::new` as public and unsafe instead.

sure, also fine with that solution -- changed